crates/viewer/re_renderer/src/video/decoder/native_av1.rs

```diff
@@ -0,0 +1,372 @@
+        // so instead return the last decoded frame.
+        if presentation_timestamp - frame.timestamp > frame.duration {
+            return Ok(VideoFrameTexture::Pending(self.texture.clone()));
+        }
+
+        if self.last_used_frame_timestamp != frame.timestamp {
+            self.last_used_frame_timestamp = frame.timestamp;
+            copy_video_frame_to_texture(&self.queue, frame, &self.texture.texture)?;
+        }
+
+        Ok(VideoFrameTexture::Ready(self.texture.clone()))
+    }
+}
+
+impl Av1VideoDecoder {
+    pub fn duration_ms(&self) -> f64 {
+        self.data.duration_sec()
+    }
+
+    pub fn width(&self) -> u32 {
+        self.data.config.coded_width as u32
+    }
+
+    pub fn height(&self) -> u32 {
+        self.data.config.coded_height as u32
+    }
+
+    /// Enqueue all samples in the given segment.
+    ///
+    /// Does nothing if the index is out of bounds.
+    fn enqueue_segment(&self, segment_idx: usize) -> Result<(), DecodingError> {
+        let Some(segment) = self.data.segments.get(segment_idx) else {
+            return Ok(());
+        };
+
+        let samples = &self.data.samples[segment.range()];
+
+        // The first sample in a segment is always a key frame:
+        self.enqueue_sample(&samples[0], true)?;
+        for sample in &samples[1..] {
+            self.enqueue_sample(sample, false)?;
+        }
+
+        Ok(())
+    }
+
+    /// Enqueue the given sample.
+    fn enqueue_sample(&self, sample: &re_video::Sample, is_key: bool) -> Result<(), DecodingError> {
+        let chunk = self.data.get(sample).ok_or(DecodingError::BadData)?;
+        self.decoder.decode(chunk);
+        Ok(())
+    }
+
+    /// Reset the video decoder and discard all frames.
+    fn reset(&mut self) {
+        re_log::debug!("Resetting AV1 decoder");
+        self.decoder.reset();
+
+        let mut decoder_output = self.decoder_output.lock();
+        decoder_output.reset_since_last_reported_error = true;
+        decoder_output.frames.clear();
+
+        self.current_segment_idx = usize::MAX;
+        self.current_sample_idx = usize::MAX;
+    }
+}
+
+fn copy_video_frame_to_texture(
+    queue: &wgpu::Queue,
+    frame: &Frame,
+    texture: &wgpu::Texture,
+) -> Result<(), DecodingError> {
+    let size = wgpu::Extent3d {
+        width: frame.width,
+        height: frame.height,
+        depth_or_array_layers: 1,
+    };
+
+    let format = match frame.format {
+        re_video::PixelFormat::Rgba8Unorm => wgpu::TextureFormat::Rgba8Unorm,
+    };
+
+    let width_blocks = frame.width / format.block_dimensions().0;
+
+    #[allow(clippy::unwrap_used)] // block_copy_size can only fail for weird compressed formats
+    let block_size = format
+        .block_copy_size(Some(wgpu::TextureAspect::All))
+        .unwrap();
+
+    let bytes_per_row_unaligned = width_blocks * block_size;
+
+    re_tracing::profile_scope!("write_texture");
+    queue.write_texture(
```

Comment:
let's not use queue here but instead our own write belt which can be a lot more efficient
(especially since it allows us to move towards writing the bytes directly into a buffer that is read by gpu, meaning we have a copy less)

Reply:
Maybe we can do that in a follow-up PR, e.g. at the same time as when we move the YUV decoding to GPU?

Reply:
Added to https://github.com/rerun-io/rerun/issues/7605